Fails to encrypt using pubkey on HSM

**Describe the bug**

Fails to encrypt (RSA OAEP) using public key on HSM device:
```
$ openssl pkeyutl -encrypt -pubin -inkey "pkcs11:model=YubiHSM;token=YubiHSM;id=%04%02;object=RSA-OAEP;type=public" -pkeyopt rsa_padding_mode:oaep -pkeyopt rsa_oaep_md:sha384 -pkeyopt rsa_mgf1_md:sha384 -in /tmp/key.bin -out /tmp/key.enc1
Enter pass phrase for PKCS#11 Token (Slot 0 - YubiHSM Connector 0.0.0.0):
Public Key operation error
8096FA56F87F0000:error:42000063:pkcs11:p11prov_EncryptInit:The specified key is not the correct type of key to use with the specified mechanism:./interface.gen.c:413:Error returned by C_EncryptInit
```

**To Reproduce**
```
$ openssl rand -out /tmp/key.bin 32
$ openssl pkeyutl -encrypt -pubin -inkey "pkcs11:model=YubiHSM;token=YubiHSM;id=%04%02;object=RSA-OAEP;type=public" -pkeyopt rsa_padding_mode:oaep -pkeyopt rsa_oaep_md:sha384 -pkeyopt rsa_mgf1_md:sha384 -in /tmp/key.bin -out /tmp/key.enc1
```
Observe the result.

**Expected behavior**

Sucessful encryption
```
$ openssl pkeyutl -engine pkcs11 -keyform engine -encrypt -pubin -inkey "pkcs11:model=YubiHSM;token=YubiHSM;id=%04%02;object=RSA-OAEP;type=public" -pkeyopt rsa_padding_mode:oaep -pkeyopt rsa_oaep_md:sha384 -pkeyopt rsa_mgf1_md:sha384 -in /tmp/key.bin -out /tmp/key.enc1
Engine "pkcs11" set.
Enter PKCS#11 token PIN for YubiHSM:
$ ll /tmp/key.enc1
-rw-r--r--  1 ur20980  wheel  384 Feb 15 12:04 /tmp/key.enc1
$ 
```

**Operating environment (please complete the following information):**
 - OS: MacOS Ventura
 - Version: 13.2.1

**Token and application used (please complete the following information):**
 - Device: YubiHSM2
 - PKCS11: YubiHSM SDK 2.4.0
 - Application: OpenSSL
 - Version: 3.0.8

**Additional context**

None.

--- a/src/keymgmt.c
+++ b/src/keymgmt.c
@@ -588,6 +588,7 @@ static int p11prov_rsa_export(void *keydata, int selection,
 {
     P11PROV_OBJ *key = (P11PROV_OBJ *)keydata;
     P11PROV_CTX *ctx = p11prov_obj_get_prov_ctx(key);
+    CK_OBJECT_CLASS class = p11prov_obj_get_class(key);
 
     P11PROV_debug("rsa export %p", keydata);
 
@@ -600,7 +601,7 @@ static int p11prov_rsa_export(void *keydata, int selection,
     }
 
     /* if anything else is asked for we can't provide it, so be strict */
-    if ((selection & ~(PUBLIC_PARAMS)) == 0) {
+    if ((class == CKO_PUBLIC_KEY) || (selection & ~(PUBLIC_PARAMS)) == 0) {
         return p11prov_obj_export_public_rsa_key(key, cb_fn, cb_arg);
     }
 
@@ -1015,6 +1016,7 @@ static int p11prov_ec_export(void *keydata, int selection, OSSL_CALLBACK *cb_fn,
 {
     P11PROV_OBJ *key = (P11PROV_OBJ *)keydata;
     P11PROV_CTX *ctx = p11prov_obj_get_prov_ctx(key);
+    CK_OBJECT_CLASS class = p11prov_obj_get_class(key);
 
     P11PROV_debug("ec export %p", keydata);
 
@@ -1027,7 +1029,7 @@ static int p11prov_ec_export(void *keydata, int selection, OSSL_CALLBACK *cb_fn,
     }
 
     /* this will return the public EC_POINT as well as DOMAIN_PARAMTERS */
-    if ((selection & ~(PUBLIC_PARAMS)) == 0) {
+    if ((class == CKO_PUBLIC_KEY) || (selection & ~(PUBLIC_PARAMS)) == 0) {
         return p11prov_obj_export_public_ec_key(key, cb_fn, cb_arg);
     }
 
@@ -1279,6 +1281,7 @@ static int p11prov_ed_export(void *keydata, int selection, OSSL_CALLBACK *cb_fn,
 {
     P11PROV_OBJ *key = (P11PROV_OBJ *)keydata;
     P11PROV_CTX *ctx = p11prov_obj_get_prov_ctx(key);
+    CK_OBJECT_CLASS class = p11prov_obj_get_class(key);
 
     P11PROV_debug("ed export %p", keydata);
 
@@ -1291,7 +1294,7 @@ static int p11prov_ed_export(void *keydata, int selection, OSSL_CALLBACK *cb_fn,
     }
 
     /* this will return the public EC_POINT */
-    if ((selection & ~(PUBLIC_PARAMS)) == 0) {
+    if ((class == CKO_PUBLIC_KEY) || (selection & ~(PUBLIC_PARAMS)) == 0) {
         return p11prov_obj_export_public_ec_key(key, cb_fn, cb_arg);
     }
 

